### ISSUE TITLE
[BUG] Build secrets from env vars loaded from .env file are empty

### ISSUE THREAD

**Issue by pschoen-itsc** [docker/compose#12037]
### Description

When using a docker compose file with secrets, where the secret value comes from an env var and this env var is only set via an .env file, then the mounted secret is empty only during build time.
What works is:
- Loading build secret from file
- Using env secret from .env during container runtime
- Setting env var used by build secret explicitly

### Steps To Reproduce

1. Create docker-compose file with a secret, which value is loaded from an env var
2. Use this secret during build time of the image
3. Define the env var in the .env file, do not set it manually
4. Execute `cat /run/secrets/<secret-name>`. The file is empty

### Compose Version

```Text
Docker Compose version v2.28.1-desktop.1
```

### Docker Environment

```Text
Client:
 Version:    27.0.3
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.15.1-desktop.1
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.28.1-desktop.1
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-compose
  debug: Get a shell into any image or container (Docker Inc.)
    Version:  0.0.32
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-debug
  desktop: Docker Desktop commands (Alpha) (Docker Inc.)
    Version:  v0.0.14
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-desktop
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.2
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.25
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-extension
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.5
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-feedback
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.3.0
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-sbom
  scout: Docker Scout (Docker Inc.)
    Version:  v1.10.0
    Path:     /Users/patrickschonthaler/.docker/cli-plugins/docker-scout

Server:
 Containers: 4
  Running: 2
  Paused: 0
  Stopped: 2
 Images: 3
 Server Version: 27.0.3
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae71819c4f5e67bb4d5ae76a6b735f29cc25774e
 runc version: v1.1.13-0-g58aa920
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
  cgroupns
 Kernel Version: 6.6.32-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 8
 Total Memory: 7.657GiB
 Name: docker-desktop
 ID: 31d53c25-4252-4a67-bc0f-fa739a4d2208
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Labels:
  com.docker.desktop.address=unix:///Users/patrickschonthaler/Library/Containers/com.docker.docker/Data/docker-cli.sock
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: daemon is not using the default seccomp profile
```

### Anything else?

Environment is Docker Desktop on Mac. Did a fresh install after getting the problem. Definitely worked before, used this construct for over a year now for my builds. Can't say when it stopped working exactly. This build step is cached most of the time.

**Comment by ndeloof** [comment]
Already fixed in 2.29.1 AFAICT, can you please give it a try ?

**Comment by pschoen-itsc** [comment]
@ndeloof Thank you. It seems to work. Can you tell me which versions are affected by this bug?
